### PR TITLE
Prepare History trie in Next RSpace for partial trie computation.

### PR DESCRIPTION
**Pull request by dzajkowski**
## Overview
<sup>_What this PR does, and why it's needed_</sup>
The structure of History trie is multilayered. This PR makes it easier to reason about changes in particular paths.

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>
RCHAIN-3415

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [ ] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by dzajkowski** [comment]
bors try

**Comment by dzajkowski** [comment]
bors try-

**Comment by dzajkowski** [comment]
bors try

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala [hunk 1]
```diff
@@ -58,9 +82,32 @@ final case class PointerBlock private (toVector: Vector[TriePointer]) extends No
...
+  @transient
+  private[this] var _hash: Blake2b256Hash = null
+
+  private[this] def calculateHash() = {
+    _data = PointerBlock.codecPointerBlock.encode(this).get
+    _hash = Blake2b256Hash.create(_data.toByteVector)
+  }
+
+  def hash: Blake2b256Hash = {
+    if (_data == null) {
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala]: `if(_hash == null)`

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala [hunk 2]
```diff
@@ -46,8 +46,32 @@ sealed trait Trie
...
+  @transient
+  private[this] var _hash: Blake2b256Hash = null
+
+  private[this] def calculateHash() = {
+    _data = Trie.codecSkip.encode(this).get
+    _hash = Blake2b256Hash.create(_data.toByteVector)
+  }
+
+  def hash: Blake2b256Hash = {
+    if (_data == null) {
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala]: And here as well

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala [hunk 3]
```diff
@@ -46,8 +46,32 @@ sealed trait Trie
 sealed trait NonEmptyTrie extends Trie
 case object EmptyTrie     extends Trie
 final case class Skip(affix: ByteVector, ptr: ValuePointer) extends NonEmptyTrie {
-  def hash: Blake2b256Hash = Blake2b256Hash.create(encode.toByteVector)
-  def encode: BitVector    = codecSkip.encode(this).get
+  @SuppressWarnings(Array("org.wartremover.warts.Var"))
+  @transient
+  private[this] var _data: BitVector = null
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala]: If you used `lazy val` instead of `var` with initialization in place you could get rid of the `null`s and `SuppressWarnings`

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package coop.rchain.rspace.nextgenrspace.history
+
+import cats.effect.Sync
+import cats.implicits._
+import cats.{Applicative, FlatMap}
+import coop.rchain.rspace.Blake2b256Hash
+import coop.rchain.rspace.nextgenrspace.history.History._
+
+object HistoryInstances {
+  def simplistic[F[_]: Sync](
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: This name is a bit vague. How is this instance `simplistic`?

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala (hunk 1 above)
Comment (dzajkowski) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala]: hash depends on data being calculated. if data is there we calculate both data and hash.

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala (hunk 4 above)
Comment (dzajkowski) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: will replace with `noMerging`

**Comment by dzajkowski** [comment]
bors try

**Comment by dzajkowski** [comment]
bors try

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+              remainingPrefixNodes: Vector[Trie],
+              nextLastSeen: Trie,
+              currentAcc: List[Trie]
+              ) =>
+            (remainingPrefixNodes, nextLastSeen) match {
+              case (head :+ (pointerBlock: PointerBlock), other) =>
+                val ptr     = pointer(other)
+                val updated = pointerBlock.updated((toInt(remainingPath.last), ptr) :: Nil)
+                Applicative[F].pure(
+                  (remainingPath.init, head, updated, currentAcc :+ nextLastSeen).asLeft
```
Comment (odeac) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: You have quite a few usages of :+ which makes me worry about the complexity of the algorithm.

What's the complexity of rehash? To me it looks quadratic in the length of KeyPath. How big can the KeyPath be? Is this a concern?

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+        case TriePath(elems :+ Skip(affix, LeafPointer(_)), _, path) => // update value
+          rehash(path.dropRight(affix.size.toInt), elems, Skip(affix, newLeaf), Nil)
+        case _ => Sync[F].raiseError(MalformedTrieError)
+      }
+
+    private def rehash(
+        path: KeyPath,
+        rest: Vector[Trie],
+        lastSeen: Trie,
+        acc: List[Trie]
```
Comment (odeac) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: acc is an accumulator only inside the go function.

This outside parameter should have a different name like initialSomething

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+    ): F[List[Trie]] = {
+      type Params = (KeyPath, Vector[Trie], Trie, List[Trie])
+      def go(params: Params): F[Either[Params, List[Trie]]] =
```
Comment (odeac) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: it doesn't make sense to group the parameters in a tuple and then un-tuple them with a match. They should be simple named parameters of the go function

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+                    )
+                  )
+                  .map(_.asLeft)
+
+              case (Vector(), _) => Applicative[F].pure(currentAcc :+ nextLastSeen).map(_.asRight)
+              case _             => Sync[F].raiseError(MalformedTrieError)
+
+            }
+        }
+      FlatMap[F].tailRecM((path, rest, lastSeen, acc))(go)
```
Comment (odeac) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: can you rewrite this `rehash` function as a fold instead of using recursivity?

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala (hunk 5 above)
Comment (dzajkowski) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: the complexity of the rehash method isn't the focus of this pr (it did not change since last implementation)

having said that: https://rchain.atlassian.net/projects/RCHAIN/issues/RCHAIN-3508?

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala (hunk 8 above)
Comment (dzajkowski) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: could you elaborate on the upside? please consider that the more I diverge from the original implementation the more it becomes hard to follow

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+          val existingHead :: existingTail = affixBytes.drop(prefixPath.size)
+          val (dividedElems, dividedTopPtr) = divideSkip(
+            incomingTail,
+            incomingHead,
+            newLeaf,
+            existingTail,
+            existingHead,
+            existingPointer,
+            prefixPath
+          )
```
[the remark this answers is not in the document]
Comment (dzajkowski) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: I don't see the duplication. they are similar, but there are like 2 repeated lines.

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala (hunk 7 above)
Comment (dzajkowski) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: I don't follow, how would this be used with a `fold` or `tailRec`?

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala (hunk 8 above)
Comment (odeac) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: I understand. Nevermind.

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala (hunk 7 above)
Comment (odeac) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: Right! I missed that part

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala (hunk 8 above)
Comment (dzajkowski) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: btw I agree that there is a fold there and I will try to refactor and see if its easier to read.

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (dzajkowski) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala]: I applied some refactor to the method, hopefully that's more readable

**Comment by dzajkowski** [comment]
bors try
